### PR TITLE
json-glib: new package

### PR DESCRIPTION
**Pull request by micmac1**
From upstream description:

JSON-GLib is a library providing serialization and deserialization
support for the JavaScript Object Notation (JSON) format described by
RFC 4627.

Signed-off-by: Sebastian Kemper <sebastian_ml@gmx.net>

Maintainer: me
Compile tested: arc770, ath79, x86_64 (all master)
Run tested: ath79 19.07 using rtpengine (uses this library)

Description:

Hi all,

I'm hoping to get this into packages. I want to add rtpengine to telephony repo. This is the one dependency that is missing. json-glib is not a telephony package, hence I'd rather it gets added to packages. I would maintain it.

I checked if introducing this lib could cause build fails for any other packages. I looked at Gentoo's packages that depend on json-glib and checked which of these are in OpenWrt. The only two I found were zerotier and grilo-plugins (maintainer @flyn-org). But in zerotier I actually cannot find any mention of json-glib, so I'm wondering if that might be a mistake in Gentoo. And in grilo-plugins the parts that can use json-glib (lua and some other part I now forget) are not packaged in OpenWrt.

I compiled both full zerotier and grilo-plugins with json-glib in staging and there was no issue. So to the best of my knowledge there wouldn't be any failure introduced. But in the unexpected case that there is build failure I would fix it.

Thanks!

Best regards,
Seb

### CODE REVIEW

File: libs/json-glib/Makefile [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+PKG_BUILD_PARALLEL:=1
+PKG_INSTALL:=1
+
+PKG_LICENSE:=LGPL-2.1+
+PKG_LICENSE_FILES:=COPYING
+PKG_MAINTAINER:=Sebastian Kemper <sebastian_ml@gmx.net>
+
+PKG_BUILD_DEPENDS:=gettext glib2/host meson/host
+
+PKG_FORTIFY_SOURCE:=0
```
Comment (neheb) [inline libs/json-glib/Makefile:27]: Ehm why?

File: libs/json-glib/Makefile [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+PKG_HASH:=720c5f4379513dc11fd97dc75336eb0c0d3338c53128044d9fabec4374f4bc47
+
+PKG_BUILD_PARALLEL:=1
+PKG_INSTALL:=1
+
+PKG_LICENSE:=LGPL-2.1+
+PKG_LICENSE_FILES:=COPYING
+PKG_MAINTAINER:=Sebastian Kemper <sebastian_ml@gmx.net>
+
+PKG_BUILD_DEPENDS:=gettext glib2/host meson/host
```
Comment (neheb) [inline libs/json-glib/Makefile:25]: gettext should be handled by nls.mk I believe.

File: libs/json-glib/Makefile [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+support for the JavaScript Object Notation (JSON) format described by
+RFC 4627.
+endef
+
+MESON_ARGS += \
+	-Ddocs=false \
+	-Dintrospection=false \
+	-Dman=false
+
+# Disable installed-tests; this also indirectly removes
```
Comment (neheb) [inline libs/json-glib/Makefile:52]: Interestingly, upstream added tests as an option. No idea if it makes sense to backport.

File: libs/json-glib/Makefile [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+PKG_RELEASE:=1
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.xz
+PKG_SOURCE_URL:=https://download.gnome.org/sources/json-glib/1.4
+PKG_HASH:=720c5f4379513dc11fd97dc75336eb0c0d3338c53128044d9fabec4374f4bc47
+
+PKG_BUILD_PARALLEL:=1
+PKG_INSTALL:=1
+
+PKG_LICENSE:=LGPL-2.1+
```
Comment (neheb) [inline libs/json-glib/Makefile:21]: should be -or-later instead of +.

File: libs/json-glib/Makefile [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+PKG_NAME:=json-glib
+PKG_VERSION:=1.4.4
+PKG_RELEASE:=1
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.xz
+PKG_SOURCE_URL:=https://download.gnome.org/sources/json-glib/1.4
+PKG_HASH:=720c5f4379513dc11fd97dc75336eb0c0d3338c53128044d9fabec4374f4bc47
+
+PKG_BUILD_PARALLEL:=1
```
Comment (neheb) [inline libs/json-glib/Makefile:18]: PKG_BUILD_PARALLEL is useless on meson packages. It's only for classical make ones.

File: libs/json-glib/Makefile [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+#
+
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=json-glib
+PKG_VERSION:=1.4.4
+PKG_RELEASE:=1
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.xz
+PKG_SOURCE_URL:=https://download.gnome.org/sources/json-glib/1.4
```
Comment (neheb) [inline libs/json-glib/Makefile:15]: I think the @GNOME macro can be used here.

**Comment by micmac1** [comment]
>+PKG_FORTIFY_SOURCE:=0
>
>Ehm why?

Bombs otherwise.

**Comment by neheb** [comment]
I'll test when I can. Format errors should be fixed IMO.

```
git grep PKG_FORTIFY_SOURCE | wc -l
2
```

I'd like to get that to 0.

**Comment by micmac1** [comment]
>+PKG_BUILD_DEPENDS:=gettext glib2/host meson/host
>
>gettext should be handled by nls.mk I believe.


Good point, will remove it.

File: libs/json-glib/Makefile (hunk 3 above)
Comment (micmac1) [reply libs/json-glib/Makefile:51]: It's already disabled with sed script.

File: libs/json-glib/Makefile (hunk 4 above)
Comment (micmac1) [reply libs/json-glib/Makefile]: changed

File: libs/json-glib/Makefile (hunk 5 above)
Comment (micmac1) [reply libs/json-glib/Makefile]: OK, removed.

File: libs/json-glib/Makefile (hunk 6 above)
Comment (micmac1) [reply libs/json-glib/Makefile]: That works, thanks. Changed.

**Comment by neheb** [comment]
I can't replicate the format errors. Is this on 64-bit?

**Comment by neheb** [comment]
Nope. Aarch64 build succeeds.

**Comment by micmac1** [comment]
>I can't replicate the format errors. Is this on 64-bit?

My laptop is 64 bits if that's what you mean.

**Comment by neheb** [comment]
I meant the target. I removed the format variable and compiled on 32 and 64-bit. Both worked.

**Comment by micmac1** [comment]
> I meant the target. I removed the format variable and compiled on 32
> and 64-bit. Both worked.

Here without PKG_FORTIFY_SOURCE:=0 it fails on both 32 and 64 bit:

In file included from ../json-glib/json-scanner.c:29:
/home/sk/tmp/openwrt/staging_dir/toolchain-mips_24kc_gcc-8.4.0_musl/include/fortify/stdio.h: In function 'snprintf':
/home/sk/tmp/openwrt/staging_dir/toolchain-mips_24kc_gcc-8.4.0_musl/include/fortify/stdio.h:101:2: error: format not a string literal, argument types not checked [-Werror=format-nonliteral]
  return __orig_snprintf(__s, __n, __f, __builtin_va_arg_pack());
  ^~~~~~
/home/sk/tmp/openwrt/staging_dir/toolchain-mips_24kc_gcc-8.4.0_musl/include/fortify/stdio.h: In function 'sprintf':
/home/sk/tmp/openwrt/staging_dir/toolchain-mips_24kc_gcc-8.4.0_musl/include/fortify/stdio.h:110:3: error: format not a string literal, argument types not checked [-Werror=format-nonliteral]
   __r = __orig_snprintf(__s, __b, __f, __builtin_va_arg_pack());
   ^~~
/home/sk/tmp/openwrt/staging_dir/toolchain-mips_24kc_gcc-8.4.0_musl/include/fortify/stdio.h:114:3: error: format not a string literal, argument types not checked [-Werror=format-nonliteral]
   __r = __orig_sprintf(__s, __f, __builtin_va_arg_pack());
   ^~~
cc1: some warnings being treated as errors

64 bit target:

In file included from ../json-glib/json-gvariant.c:22:
/home/sk/tmp/sdk/master-x86_64/staging_dir/toolchain-x86_64_gcc-8.4.0_musl/include/fortify/stdio.h: In function 'snprintf':
/home/sk/tmp/sdk/master-x86_64/staging_dir/toolchain-x86_64_gcc-8.4.0_musl/include/fortify/stdio.h:101:2: error: format not a string literal, argument types not checked [-Werror=format-nonliteral]
  return __orig_snprintf(__s, __n, __f, __builtin_va_arg_pack());
  ^~~~~~
/home/sk/tmp/sdk/master-x86_64/staging_dir/toolchain-x86_64_gcc-8.4.0_musl/include/fortify/stdio.h: In function 'sprintf':
/home/sk/tmp/sdk/master-x86_64/staging_dir/toolchain-x86_64_gcc-8.4.0_musl/include/fortify/stdio.h:110:3: error: format not a string literal, argument types not checked [-Werror=format-nonliteral]
   __r = __orig_snprintf(__s, __b, __f, __builtin_va_arg_pack());
   ^~~
/home/sk/tmp/sdk/master-x86_64/staging_dir/toolchain-x86_64_gcc-8.4.0_musl/include/fortify/stdio.h:114:3: error: format not a string literal, argument types not checked [-Werror=format-nonliteral]
   __r = __orig_sprintf(__s, __f, __builtin_va_arg_pack());
   ^~~
cc1: some warnings being treated as errors

**Comment by neheb** [comment]
Aha. That makes sense. I have fortify-source disabled locally.

I will try this on my VM and see if I can do anything.

**Comment by micmac1** [comment]
Mmh, that looks more like an issue with glib, actually. Do you compile against the new glib version for which you have a pull request open? I'm using 2.58.3 like it is in master right now.

**Comment by neheb** [comment]
I don't see it as a glib issue. I also don't think the version matters. But yes, I use 2.64.2 locally as well.

**Comment by neheb** [comment]
Reproduced. Looks like a garbage fortify-source failure. No idea how to fix.

**Comment by neheb** [comment]
Found a way to fix is. Remove error=format=2 from meson.build.

**Comment by neheb** [comment]
Is nls.mk needed?

**Comment by micmac1** [comment]
> Found a way to fix is. Remove error=format=2 from meson.build.

OK, I removed that and PKG_FORTIFY_SOURCE:=0. Thanks.

**Comment by micmac1** [comment]
> Is nls.mk needed?

Yes, otherwise glib2 might not find libintl.h. Saw that happening on ARC with uclibc.

**Comment by neheb** [comment]
Hmm OK. glib2 2.64.2 will remove this requirement, but for now it's fine.

File: libs/json-glib/Makefile [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+# Disable installed-tests; this also indirectly removes
+# build_aux/gen-installed-test.py calls (copied from Gentoo)
+define Build/Prepare
+	$(call Build/Prepare/Default)
+	$(SED) 's/install: true/install: false/g' \
+		$(PKG_BUILD_DIR)/json-glib/tests/meson.build
+	$(SED) '/install_data/d' \
+		$(PKG_BUILD_DIR)/json-glib/tests/meson.build
+	$(SED) '/error=format=2/d' $(PKG_BUILD_DIR)/meson.build
```
Comment (neheb) [inline libs/json-glib/Makefile:59]: :). I like this better than actually patching the file.

File: libs/json-glib/Makefile [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+PKG_INSTALL:=1
+
+PKG_LICENSE:=LGPL-2.1-or-later
+PKG_LICENSE_FILES:=COPYING
+PKG_MAINTAINER:=Sebastian Kemper <sebastian_ml@gmx.net>
+
+PKG_BUILD_DEPENDS:=glib2/host meson/host
+
+OPENWRT_VERBOSE:=c
```
Comment (neheb) [inline libs/json-glib/Makefile:26]: Why is this here?

**Comment by micmac1** [comment]
>+OPENWRT_VERBOSE:=c
>
>Why is this here?

I like seeing the compiler calls when something goes south.
